torch/_dynamo/variables/torch.py

```diff
@@ -789,20 +790,30 @@ def speculate_subgraph(
             args = []
             # One argument to graph per sub_args
             for a in sub_args:
-                if a is None:
-                    new_arg = ConstantVariable(None)
                 assert not isinstance(
                     a, torch.Tensor
                 ), "Tensors should already be tracked?"
+                if a is None:
+                    a = ConstantVariable(None)
+
                 if isinstance(a, ConstantVariable):
-                    proxy = tracer.create_graph_input("const")
+                    # This arg is not used in the body of the higher order op.
+                    # Currently, this new input is added to make the calls
+                    # happy, which expect a fixed number of arguments. In
+                    # future, we can clean this up.
+                    tracer.create_graph_input("const")
+                    # Ensures that we recompile when the constant value changes
+                    a.add_guard(GuardBuilder.CONSTANT_MATCH)
```

Comment:
One thing I am curious about is: why doesn't a guard on the constant get added when dynamo sees it being used?

In the following test case,
```py
        x = torch.randn(3, 3)
        y = 4.0

        def fn(x, y, z):
            if z:
                return x + y
            return x * y

        def f(x, y, z):
            return wrap(fn, x, y, z)

        args = (x, 4.0, None)
        opt_f = torch.compile(f, fullgraph=True, backend=CompileCounter())
```
I would have naively expected dynamo to add a guard when it sees `x * y` and notices that `y` is a constant

Reply:
That is a good question! 
Yeah, I don't know. Let's revisit this. I am going to note this down in tracker.